daliuge-translator/dlg/dropmake/cwl.py

```diff
@@ -0,0 +1,148 @@
+
+#    ICRAR - International Centre for Radio Astronomy Research
+#    (c) UWA - The University of Western Australia, 2015
+#    Copyright by UWA (in the framework of the ICRAR)
+#    All rights reserved
+#
+#    This library is free software; you can redistribute it and/or
+#    modify it under the terms of the GNU Lesser General Public
+#    License as published by the Free Software Foundation; either
+#    version 2.1 of the License, or (at your option) any later version.
+#
+#    This library is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+#    Lesser General Public License for more details.
+#
+#    You should have received a copy of the GNU Lesser General Public
+#    License along with this library; if not, write to the Free Software
+#    Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+#    MA 02111-1307  USA
+#
+
+import logging
+import os
+from zipfile import ZipFile
+
+import cwlgen
+
+from dlg import common
+
+
+#from ..common import dropdict, get_roots
+logger = logging.getLogger(__name__)
+
+
+def create_workflow(pgt, pgt_path, cwl_path, zip_path):
+    """
+    Create a CWL workflow from a given Physical Graph Template
+
+    A CWL workflow consists of multiple files. A single file describing the
+    workflow, and multiple files each describing one step in the workflow. All
+    the files are combined into one zip file, so that a single file can be
+    downloaded by the user.
+
+    NOTE: CWL only supports workflow steps that are bash shell applications
+          Non-BashShellApp nodes are unable to be implemented in CWL
+    """
+
+    # create list for command line tool description files
+    step_files = []
+
+    # create the workflow
+    cwl_workflow = cwlgen.Workflow('', label='', doc='', cwl_version='v1.0')
+
+    # create files dictionary
+    files = {}
+
+    # look for input and output files in the pg_spec
+    for index, node in enumerate(pgt):
+        command = node.get('command', None)
+        dataType = node.get('dt', None)
+        outputId = node.get('oid', None)
+        outputs = node.get('outputs', [])
+
+        if len(outputs) > 0:
+            files[outputs[0]] = "step" + str(index) + "/output_file_0"
+
+    # add steps to the workflow
+    for index, node in enumerate(pgt):
+        dataType = node.get('dt', '')
+
+        if dataType == 'BashShellApp':
```

Comment:
I'm assuming the current behavior when we try to translate a graph that doesn't have bash apps is that we get an empty workflow? We could consider throwing an error instead, as the workflow returning from here is potentially a subset of the input PG, and users might not be quite aware of this. I'll add something into our backlog for this.

Reply:
Yes, we'd have to add the ability to notify the user of this issue in both cases:

- When they run dlg cwl on the command line
- When they click the "Export to CWL" button on the HTML page served by lg web.